File: Typescript/src/DelegationSortedIntList.ts
/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */

import { SortedIntList } from "./hidden/SortedIntListLibrary"
import { IntegerList } from "./IntegerList"

class DelegationSortedIntList implements IntegerList{

  private sortedIntList: SortedIntList
  private totalAdded: number


  constructor() {
    this.sortedIntList = new SortedIntList()
    this.totalAdded = 0
  }

  add (num: number) : boolean {
    this.totalAdded += 1;
    return this.sortedIntList.add(num)
  }

  addAll (list: IntegerList) : boolean {
    this.totalAdded += list.size()
    return this.sortedIntList.addAll(list)
  } 

  get (index: number) : number {
    return this.sortedIntList.get(index)
  }

  remove (num: number) : boolean {
    return this.sortedIntList.remove(num)
  }

  removeAll (list: IntegerList) : boolean {
    return this.sortedIntList.removeAll(list)
  }

  size () : number {
    return this.sortedIntList.size()
  }

  getTotalAdded () : number {
    return this.totalAdded
  }
}

export { DelegationSortedIntList }

File: Java/src/main/java/edu/cmu/cs/cs214/rec04/DelegationSortedIntList.java
package edu.cmu.cs.cs214.rec04;

/**
 * DelegationSortedIntList -- a variant of a SortedIntList that keeps
 * count of the number of attempted element insertions (not to be confused
 * with the current size, which goes down when an element is removed)
 * and exports an accessor (totalAdded) for this count.
 *
 * @author Nora Shoemaker
 *
 */

public class DelegationSortedIntList implements IntegerList{
    private int totalAdded;
    private SortedIntList sortedIntList;

    public DelegationSortedIntList() {
        this.totalAdded = 0;
        this.sortedIntList = new SortedIntList();
    }

    public boolean add(int num){
        totalAdded += 1;
        return sortedIntList.add(num);
    }

    public boolean addAll(IntegerList list){
        totalAdded += list.size();
        return sortedIntList.addAll(list);
    }

    public int get(int index){
        return sortedIntList.get(index);
    }

    public boolean remove(int num){
        return sortedIntList.remove(num);
    }

    public boolean removeAll(IntegerList list) {
        return sortedIntList.removeAll(list);
    }

    public int size() {
        return sortedIntList.size();
    }
    
    public int getTotalAdded() {
        return totalAdded;
    }
    
}